File: 0046-permutations/0046-permutations.java
class Solution {
    List<List<Integer>>ans = new ArrayList<>();
    public List<List<Integer>> permute(int[] nums) {
        List<Integer>list = new ArrayList<>();
        backtrack(list, nums, 0);
        return ans;
    }
    private void backtrack(List<Integer>list, int[] nums, int index){
        if(index==nums.length){
            ans.add(new ArrayList<>(list));
        }else{
            for(int i=0;i<nums.length;i++){
                if(list.contains(nums[i])) continue;
                list.add(nums[i]);
                backtrack(list, nums, index+1);
                list.remove(list.size()-1);
            }
        }
    }
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
    vector<vector<int>>ans;
    vector<vector<int>> permute(vector<int>& nums) {
        vector<int>temp;
        vector<bool> used(nums.size(), false);
        recur(nums, 0, temp, used);
        return ans;
    }
    void recur(vector<int>& nums, int index, vector<int>&temp,vector<bool>&used){
        //ans.push_back(temp);
        if(index == nums.size()){
            ans.push_back(temp);
        }else{
            for(int i=0;i<nums.size();i++){
                if(!used[i]){
                    used[i]=true;
                    temp.push_back(nums[i]);
                    recur(nums, index+1, temp,used);
                    temp.pop_back();
                    used[i]=false;
                }
            }
        }
        
    }
};